### PR TITLE
[Spec Decode] Avoid materializing target probabilities in rejection sampling

### PR DESCRIPTION
**Pull request by masterFoad**
## Summary

This removes the materialized `target_probs = softmax(target_logits)` tensor from the non-greedy speculative rejection sampler.

The sampler still uses the same softmax probabilities. Instead of writing the full probability tensor, it computes one logsumexp value per row and reconstructs probabilities where they are used:

```python
prob = exp(logit - logsumexp(row))
```

## Why

The old path allocates and reads a full fp32 probability tensor:

```text
num_draft_tokens_total x vocab_size x 4 bytes
```

At batch 64, k=16, that transient tensor is:

| vocab | removed transient tensor |
|---:|---:|
| 50,272 | 205.9 MB |
| 128,256 | 525.3 MB |
| 151,936 | 622.3 MB |

This reduces transient GPU memory pressure and HBM traffic in the speculative rejection path.

This is independent of #41258: #41258 removes eager recovery buffers, while this PR removes the materialized target probability tensor. Both touch the rejection sampler, so whichever lands first, the other will rebase. Together, the two changes remove two separate full-vocab tensors from this path.

## Changes

- Added a one-pass Triton logsumexp producer for target logits.
- Passed `target_logits` and per-row LSE into the rejection kernels instead of `target_probs`.
- Reconstructed target probabilities at point of use with `exp(logit - lse)`.
- Kept the same acceptance and recovery sampling algorithms.
- Added coverage for LSE parity, fp64-reference accuracy, and fully masked `-inf` vocab tiles.

## Correctness

The probability identity is exact in real arithmetic:

```text
softmax(logits)[i] == exp(logits[i] - logsumexp(logits))
```

Floating point reduction order can differ from materialized softmax by a few ULPs. The tests check acceptance parity outside threshold-edge cases and compare the fused LSE producer against an fp64 reference.

## Performance

A100 mechanism benchmark, paired against latest main, 200 trials per row:

| vocab | draft probs | reject pos | main median | this PR median | delta |
|---:|---|---:|---:|---:|---:|
| 50,272 | false | none | 0.9099 ms | 0.6457 ms | -29.03% |
| 50,272 | false | 0 | 0.9484 ms | 0.6351 ms | -33.04% |
| 50,272 | true | 0 | 0.9157 ms | 0.8422 ms | -8.02% |
| 128,256 | false | none | 2.1401 ms | 1.4180 ms | -33.74% |
| 128,256 | false | 15 | 2.1218 ms | 1.2975 ms | -38.85% |
| 128,256 | true | 0 | 2.1406 ms | 1.9066 ms | -10.93% |

Full sweep: faster in all 20 probed rows, with median mechanism latency reduced by 6.97% to 38.85%. The equivalent mechanism throughput gain is 7.49% to 63.53%.

No end-to-end throughput claim is made in this PR.

## Duplicate-work check

I searched open vLLM PRs for `target_probs rejection_sampler`, `compute_target_lse`, `logsumexp rejection sampler`, and `target probabilities rejection sampling`. I did not find another PR removing `target_probs` from this path.

Related open work is different in scope: #41258 makes recovery sampling lazy, #45060 fixes an out-of-vocab recovered-token bug, and #45229 changes relaxed acceptance behavior.

## Test plan

Local checks:

```bash
uv run --no-project python -m py_compile vllm/v1/sample/rejection_sampler.py tests/v1/sample/test_rejection_sampler.py
uvx ruff check vllm/v1/sample/rejection_sampler.py tests/v1/sample/test_rejection_sampler.py
```

A100 validation:

```text
vllm-lse-clean-validate-20260612t200123z
NVIDIA A100-SXM4-80GB, driver 550.54.14
ruff passed
focused full -inf tile LSE test passed
focused LSE acceptance parity test passed
focused fp64 oracle test passed
tests/v1/sample/test_rejection_sampler.py: 77 passed
```

A100 mechanism benchmark:

```text
vllm-lse-clean-validate-20260612t200123z
NVIDIA A100-SXM4-80GB, driver 550.54.14
20/20 rows faster than main, 200 trials per row
```

## AI assistance disclosure

AI assistance was used. I reviewed the changed code, tests, duplicate-work check, and benchmark results before submitting.

### CODE REVIEW

**Comment by masterFoad** [comment]
I also benchmarked `torch.compile(torch.logsumexp)` on A100. It was substantially faster than eager PyTorch, but the specialized Triton kernel I wrote was still 1.8–2.2× faster across the tested masked-logit shapes, with matching fp32-level accuracy. This supports keeping the custom online LSE kernel.
